README
This is linuxinfo, used on Linux platforms.

The latest releases are always available at:

        http://www.helgefjell.de/linuxinfoitem.php?name=download

Previously version were available at
	http://www.munted.org.uk/programming/
(Note that this domain is no longer active).

linuxinfo was initially written by Alex Buell, alex.buell@munted.org.uk.
Maintainership was transferred to Helge Kreutzmann <debian@helgefjell.de>
in 2014 who integrated the patches that had been accumulated in the Debian
versions of linuxinfo.

See the CREDITS file for a list of folks who have helped us out. linuxinfo
is copyrighted under the terms of the GPL 2+ and LGPL 2 (See the COPYING file for
the details).[1]
 
To build, simply use:

	./configure 
	make

Once it's finished the make process, su to root (if not already root), and
use:

	make install

to install the linuxinfo binary and man page to the appropriate
directories. Default is for the binary to go into /usr/local/bin and for
the man page to go into /usr/local/man/man8. You can change this by using
the --prefix in ./configure if required.

Linuxinfo contains a German translation of the man page which is currently
not installed by default.

If linuxinfo does not recognize your CPU, please send me a copy of your 
/proc/cpuinfo file, and I shall fix it for next version. Thanks. 

NOTE! linuxinfo has no capability to detect the clock speed on kernels
older than 2.1.x, *or* on low end processors.

[1] On 2014-04-26 Alex agreed to relicense linuxinfo to GPL 2+ (from GPL 2).
    However, not all contributors have yet been contacted, so the "+" is 
    pending agreement from them.

    Currently agreed: 
    Alex Buell
    Helge Kreutzmann
    Martin Michlmayr
    Andreas Jochens
    Erick Kinnee
    Ola Lundqvist


ChangeLog
05 July 2015 Helge Kreutzmann <debian@helgefjell.de>

- Release version 2.2.3

22 November 2014 Helge Kreutzmann <debian@helgefjell.de>

- The first Core Quad

12 November 2014 Helge Kreutzmann <debian@helgefjell.de>

- Add another Core 

03 November 2014 Helge Kreutzmann <debian@helgefjell.de>

- Don't add the vendor name inside CPU names, it's redundant
- Add another Core 2 Duo

12 October 2014 Helge Kreutzmann <debian@helgefjell.de>

- Add an "AMD A6-6310 with Radeon R4"
- Release version 2.2.2

30 September 2014 Helge Kreutzmann <debian@helgefjell.de>

- Fix a comment in m4/linuxinfo-progs.m4

29 September 2014 Helge Kreutzmann <debian@helgefjell.de>

- Fix compilation on PPC(64el) and S390(x).

20 September 2014 Helge Kreutzmann <debian@helgefjell.de>

- Remove useless (commented out) memory size calcuation

17 September 2014 Helge Kreutzmann <debian@helgefjell.de>

- Print "?" if no bogomips could be detected

15 September 2014 Helge Kreutzmann <debian@helgefjell.de>

- Remove useless variable tempbogo where applicable

31 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Release 2.1.0

30 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial danish man page translation. Thanks to Joe Hansen
<joedalton2@yahoo.dk>

24 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial polish man page translation. Thanks to Micha◈ Ku◈ach
<michal.kulach@gmail.com>

24 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial italian man page translation. Thanks to Beatrice Torracca
<beatricet@libero.it>

21 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial russian man page translation. Thanks to Yuri Kozlov
<yuray@komyakino.ru>

12 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial czech man page translation. Thanks to Michal Simunek
<michal.simunek@gmail.com>

10 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial portugese man page translation. Thanks to Américo Monteiro
<a_monteiro@gmx.com>

09 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Po4a infrastructure now operational for man pages

13 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Release 2.0.1

12 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial czech translation thanks to Michal Simunek
  <michal.simunek@gmail.com>

2 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Initial japanese translation thanks to victory <victory.deb@gmail.com>

1 August 2014 Helge Kreutzmann <debian@helgefjell.de>

- Release 2.0.0

31 July 2014 Helge Kreutzmann <debian@helgefjell.de>

 - Initial italian translation thanks to Beatrice Torracca
   <beatricet@libero.it>

31 July 2014 Helge Kreutzmann <debian@helgefjell.de>

 - Initial spanish translation thanks to "Matias A. Bellone"
   <matiasbellone+debian@gmail.com>

30 July 2014 Helge Kreutzmann <debian@helgefjell.de>

 - Initial polish translation thanks to Micha◈ Ku◈ach <michalkulach@gmail.com>

28 July 2014 Helge Kreutzmann <debian@helgefjell.de>

 - Initial french translation thanks to Jean-Pierre Giraud
   <jean-pierregiraud@neuf.fr>

20 July 2014 Helge Kreutzmann <debian@helgefjell.de>

 - Initial danish translation thanks to Joe Dalton <joedalton2@yahoo.dk>

17 July 2014 Helge Kreutzmann <debian@helgefjell.de>
 
 - Initial portugese translation thanks to Américo Monteiro
   <a_monteiro@gmx.com>
 - Initial russian translation thanks to Yuri Kozlov <yuray@komyakino.ru>

26 May 2014 Helge Kreutzmann <debian@helgefjell.de>

- Start relicensing from GPL 2 → GPL 2+

1 May 2014 Helge Kreutzmann <debian@helgefjell.de>

- Add initial support for arm64 (untested)

17 April 2014 Helge Kreutzmann <debian@helgefjell.de>

- Correct a printf call in linuxinfo.c

30 March 2014 Helge Kreutzmann <debian@helgefjell.de>

- Update README
- No need to have separate Changelogs in m4/ and po/
- Do not ship po/*.gmo and po/stamp-po
- Update Alex e-mail adress everywhere
- Update copyright notices in source files
- Release 1.9.4

13 March 2014 Helge Kreutzmann <debian@helgefjell.de>

- Rename linuxinfo.1.de into linuxinfo.de.1
- Add myself in AUTHOR

06 February 2014 Helge Kreutzmann <debian@helgefjell.de>

- Fix printing of Bogomips value for non C locales

06 December 2013 Helge Kreutzmann <debian@helgefjell.de>

- Autogenerate auto* files
- Makefile.am (ACLOCAL_AMFLAGS): New variable. 
  (EXTRA_DIST): Add config.rpath, m4/ChangeLog.
  from gettext-0.18.3.
- configure.ac (AC_OUTPUT): Add po/Makefile.in.
-Initial i18n by gettext, not yet 100% perfect (especially for plural forms)
-Initial German translation

14 November 2013 Helge Kreutzmann <debian@helgefjell.de>

- Add another Xeon.

28 September 2013 Helge Kreutzmann <debian@helgefjell.de>

- Add the first Athlon 62 X2
- Always parse /proc/meminfo

4 November 2011 Helge Kreutzmann <debian@helgefjell.de>

- Add another Xeon

25 July 2011 Helge Kreutzmann <debian@helgefjell.de>

- Add first Phenom II CPU

04 January 2011 Helge Kreutzmann <debian@helgefjell.de>

- Fix resource leak found by DACA/cppcheck.

10 March 2010 Nobuhiro Iwamatsu <iwamatsu@nigauri.org>

- Fix build on sh4.

28 February 2010 Helge Kreutzmann <debian@helgefjell.de>

- Add initial avr32 port.
- Add another Core Duo

1 January 2010 Nobuhiro Iwamatsu <iwamatsu@nigauri.org>

- Renesas SuperH support

5 June 2009 Helge Kreutzmann <debian@helgefjell.de>

- Add another Core Duo CPU.

24 May 2009 Helge Kreutzmann <debian@helgefjell.de>

- Add another Xeon CPU
- Relax check for /proc/meminfo usage
  On some (custom?) kernels the value used for switching to /proc/meminfo
  seems to differ. For now include the newly found values, a proper fix
  should be added later.

19 March 2009 Helge Kreutzmann <debian@helgefjell.de>

- Add Core Solo CPU.

3 February 2008 Helge Kreutzmann <debian@helgefjell.de>

- Add one Core 2 and two more Pentium 4.

10 August 2007 Helge Kreutzmann <debian@helgefjell.de>

- Add another Xeon CPU
- Don't trust /proc/kcore sizes >\approx 4 GB on x86 machines


linuxinfo_arm.c
/*
	linuxinfo_arm.c

        Copyright (C) 1998-2000
        All Rights Reserved.

        Alex Buell <alex.buell@munted.eu>

        Copyright (C) 2005
        Helge Kreutzmann <debian@helgefjell.de>

        Version Author  Date            Comments
        ----------------------------------------------------------------------
        1.0.0   AIB     199803??        Initial development
	1.0.1	AIB	20000405	Rewritten
	1.0.2	AIB	20010909	Added getphysicalmemory() call
	1.0.3	KRE	20051118	Include <string.h>

	This is the ARM port of linuxinfo
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "linuxinfo.h"

#ifdef system_arm

void GetHardwareInfo(int fd, struct hw_stat *hw)
{
	int processors = 0; 
	float bogomips = 0.0;
	float tempMHz = 0.0;

	char temp_string[BUFSIZ], temp_string2[BUFSIZ];
	char chip[BUFSIZ], vendor[BUFSIZ], model[BUFSIZ], Mhz[BUFSIZ];

	char *p;

	while (read_line(fd, temp_string, BUFSIZ) != 0)
	{
		if (splitstring(temp_string, temp_string2))
		{
			if (strncmp(temp_string, "Processor", strlen("Processor")) == 0)
			{
				if (strncmp(temp_string2, "Intel sa110", strlen("Intel sa110")) == 0)
				{
					strcpy(vendor, "Intel");
					strcpy(model, "SA110");
				}
				processors++;
			}

			if (strncmp(temp_string, "Type", strlen("Type")) == 0)
			{
				if (strncmp(temp_string2, "sa110", strlen("sa110")) == 0)
				{
					strcpy(vendor, "Intel");
					strcpy(model, "SA110");
				}
				processors++;
			}

			if (strncmp(temp_string, "Type", strlen("Type")) == 0)
				processors++;

			if (strncmp(temp_string, "cpu MHz", strlen("cpu MHz")) == 0)
				tempMHz = atol(temp_string2);

			if (strncmp(temp_string, "BogoMips", strlen("BogoMips")) == 0)
				bogomips = bogomips + atof(temp_string2);
		}

		
	}

	sprintf(hw->hw_memory, LONGSPEC, getphysicalmemory());
	hw->hw_processors = processors;

	sprintf(hw->hw_cpuinfo, "%s %s", vendor, model);
	if (bogomips == 0.0)
	    sprintf(hw->hw_bogomips, "?");
	else
	    sprintf(hw->hw_bogomips, "%0.2f", bogomips);

	if (tempMHz == 0.0)
		strcpy(hw->hw_megahertz, "?");
	else
		sprintf(hw->hw_megahertz, "%d", (int)tempMHz);
}

#endif /* system_arm */


linuxinfo_ppc.c
/*
        linuxinfo_ppc.c

        Copyright (C) 1998-2000
        All Rights Reserved.

        Alex Buell <alex.buell@munted.eu>

	Copyright (C) 2005 
	Helge Kreutzmann <debian@helgefjell.de>

        Version Author  Date            Comments
        ----------------------------------------------------------------------
        1.0.0   AIB     199803??        Initial development
	1.0.1	AIB	20000405	Rewritten
	1.0.2	AIB	20010909	Added getphysicalmemory() call
	1.0.3   KRE     20050328        Fix multicpu-detection without bogomips
	1.0.4   KRE     20051118        Include <string.h>

        This is the PPC port of linuxinfo
*/

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <string.h>

#include "linuxinfo.h"

#ifdef system_ppc

void GetHardwareInfo(int fd, struct hw_stat *hw)
{
	int processors = 0; 
	int megahertz = 0;
	float tempbogo;

	char temp_string[BUFSIZ], temp_string2[BUFSIZ];
	char chip[BUFSIZ], vendor[BUFSIZ], model[BUFSIZ];

	while (read_line(fd, temp_string, BUFSIZ) != 0)
	{
		splitstring(temp_string, temp_string2);

		//if (!(strncmp(temp_string, "cpu", strlen("cpu"))) && (processors == 0))
		if (!(strncmp(temp_string, "cpu", strlen("cpu"))))
		{
			if (processors == 0) strcpy(chip, temp_string2);
			processors++;
		}

		if (!(strncmp(temp_string, "model", strlen("model"))))
			strcpy(vendor, temp_string2);

		if (!(strncmp(temp_string, "machine", strlen("machine"))))
			strcpy(vendor, temp_string2);

		if (!(strncmp(temp_string, "clock", strlen("clock"))))
			megahertz = atoi(temp_string2);

		if (!(strncmp(temp_string, "bogomips", strlen("bogomips"))))
		{
			sscanf(temp_string2, "%f", &tempbogo);
		}
	}

	sprintf(hw->hw_memory, LONGSPEC, getphysicalmemory());
	hw->hw_processors = processors;

	if (megahertz > 0)
		sprintf(hw->hw_megahertz, "%d", megahertz);
	else
		strcpy(hw->hw_megahertz, "?");

	sprintf(hw->hw_cpuinfo, "%s %s", chip, vendor);
	if (tempbogo == 0.0)
	    sprintf(hw->hw_bogomips, "?");
	else
	    sprintf(hw->hw_bogomips, "%0.2f",tempbogo * processors);

}
#endif /* system_ppc */


linuxinfo_hppa.c
/*
        linuxinfo_hppa.c

        Copyright (C) 2002
	All Rights Reserved for Modifications.
	Ola Lundqvist <opal@debian.org>

        Copyright (C) 1998-2000
        All Rights Reserved.

        Alex Buell <alex.buell@munted.eu>

        Advanced Buell Software Engineering Ltd
        Hampshire, GU31 5DG
        United Kingdom

        Copyright (C) 2005
        Helge Kreutzmann <debian@helgefjell.de>

        Version Author  Date            Comments
        ----------------------------------------------------------------------
        1.0.0   AIB     199803??        Initial development
	1.0.1	AIB	20000405	Updated and reworked code
	1.0.2	AIB	20000527	Added AMD K7
        1.0.3   OPAL    20021225	Rewritten some for hppa.
	1.0.4   KRE     20050318        Print CPU
	1.0.5   KRE     20051118        Include <string.h>

        This is the hppa port of linuxinfo
*/

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <string.h>

#include "linuxinfo.h"

#ifdef system_hppa

static char *models[] =
{
  //"9000/800/A500-5X", "Crescendo 550",
  //"9000/778/B180L", "Merlin L2+ 180",
  //"9000/785/C3600", "Allegro W+",
  //"9000/871/D270", "UL 2w U160 512K/512K D370",
  NULL, NULL
};

void GetHardwareInfo(int fd, struct hw_stat *hw)
{
	char chip[BUFSIZ], model[BUFSIZ], Mhz[BUFSIZ], family[BUFSIZ];
	char temp_string[BUFSIZ], temp_string2[BUFSIZ];

	struct stat st_buf;

	int processors = 0; 
	float bogomips = 0.0;
	float tempMHz = 0.0;
	LONGLONG memory = 0;

	char *p;

	strcpy(chip, "Unknown");
	strcpy(model, "Unknown");
	strcpy(family, "Unknown");

	while (read_line(fd, temp_string, BUFSIZ) != 0) 
	{
		if (splitstring(temp_string, temp_string2))
		{
			if (strncasecmp(temp_string, "processor", strlen("processor")) == 0)
				processors++;

			// DEBUG printf("%s:\n", temp_string);
//			if (strncasecmp(temp_string, "model", strlen("model")) == 0) 
//			{
//				int i = 0;
//		
//                                if (strncmp(temp_string2, "0", 1) == 0)
//                                {
//                                    strcpy(model, family);
//                                }
//                                else
//                                {
//                                        while (models[i] != NULL)
//                                        {
//					        if (strncasecmp(temp_string2, models[i], strlen(models[i])) == 0)
//						        strcpy(model, models[++i]);
//
//					        i += 2;
//                                        }
//				}
//			}

			if (strncasecmp(temp_string, "family", strlen("cpu family")) == 0)
				strcpy(family, temp_string2);

			if (strncasecmp(temp_string, "cpu  ", strlen("cpu  ")) == 0)
			{
				strcpy(chip, temp_string2);
				strcpy(model, temp_string2);
			}

			if (strncasecmp(temp_string, "cpu MHz", strlen("cpu MHz")) == 0)
				tempMHz = atol(temp_string2);

			if (strncasecmp(temp_string, "bogomips", strlen("bogomips")) == 0)
				bogomips = bogomips + atof(temp_string2);
		}

		
	}

	
	sprintf(hw->hw_memory, LONGSPEC, getphysicalmemory());

	hw->hw_processors = processors;
	sprintf(hw->hw_cpuinfo, "%s", model);

	if (bogomips == 0.0)
	    sprintf(hw->hw_bogomips, "?");
	else
	    sprintf(hw->hw_bogomips, "%0.2f", bogomips);

	if (tempMHz == 0.0)
		sprintf(hw->hw_megahertz, "?");
	else
		sprintf(hw->hw_megahertz, "%d", (int)tempMHz);
}

#endif /* system_hppa */


linuxinfo_ia64.c
/*
        linuxinfo_ia64.c

        Copyright (C) 1998-2000
        All Rights Reserved.

        Alex Buell <alex.buell@munted.eu>

        Advanced Buell Software Engineering Ltd
        Hampshire, GU31 5DG
        United Kingdom

	Copyright (C) 2002 
	John R. Daily <jdaily@progeny.com>

	Copyright (C) 2005 
	Helge Kreutzmann <debian@helgefjell.de>

        Version Author  Date            Comments
        ----------------------------------------------------------------------
        1.0.0   AIB     199803??        Initial development
	1.0.1	AIB	20000405	Updated and reworked code
	1.0.2	AIB	20000527	Addded AMD K7
        1.0.3   OPAL    20030417	Fixed so that it compiles.
	1.0.4   KRE     20050318        For now always use family as model (improve later)
	1.0.5   KRE     20051118        Include <string.h>

        This is the Intel ia64 port of linuxinfo
*/

#include <stdio.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <string.h>

#include "linuxinfo.h"

#ifdef system_ia64

static char *vendors[] =
{
	"GenuineIntel", "Intel",
	NULL, NULL
};

static char *models[] =
{
        NULL, NULL
};

void GetHardwareInfo(int fd, struct hw_stat *hw)
{
	char chip[BUFSIZ], vendor[BUFSIZ], model[BUFSIZ], Mhz[BUFSIZ], family[BUFSIZ];
	char temp_string[BUFSIZ], temp_string2[BUFSIZ];

	struct stat st_buf;

	int processors = 0; 
	float bogomips = 0.0;
	float tempMHz = 0.0;
	LONGLONG memory = 0;

	char *p;

	strcpy(chip, "Unknown");
	strcpy(vendor, "Unknown");
	strcpy(model, "Unknown");
	strcpy(family, "Unknown");

	while (read_line(fd, temp_string, BUFSIZ) != 0) 
	{
		if (splitstring(temp_string, temp_string2))
		{
			if (strncasecmp(temp_string, "processor", strlen("processor")) == 0)
				processors++;

			if (strncasecmp(temp_string, "vendor", strlen("vendor")) == 0)
			{
				int i = 0;

				while (vendors[i] != NULL)
				{
					if (strncasecmp(temp_string2, vendors[i], strlen(vendors[i])) == 0)
						strcpy(vendor, vendors[++i]);

					i += 2;
				}
			}

			if (strncasecmp(temp_string, "model", strlen("model")) == 0) 
			{
				int i = 0;
		
                                //if (strncmp(temp_string2, "0", 1) == 0)
                                if (1)
                                {
                                    strcpy(model, family);
//				    printf("Model: %s\n",family);
                                }
                                else
                                {
                                        while (models[i] != NULL)
                                        {
					        if (strncasecmp(temp_string2, models[i], strlen(models[i])) == 0)
						        strcpy(model, models[++i]);

						printf("Model (m): %s\n",model);
					        i += 2;
                                        }
				}
			}

			if (strncasecmp(temp_string, "family", strlen("family")) == 0)
				strcpy(family, temp_string2);

			if (strncasecmp(temp_string, "cpu MHz", strlen("cpu MHz")) == 0)
				tempMHz = atol(temp_string2);

			if (strncasecmp(temp_string, "bogomips", strlen("bogomips")) == 0)
				bogomips = bogomips + atof(temp_string2);
		}

		
	}

        sprintf(hw->hw_memory, LONGSPEC, getphysicalmemory());
	
	hw->hw_processors = processors;
	sprintf(hw->hw_cpuinfo, "%s %s", vendor, model);

	if (bogomips == 0.0)
	    sprintf(hw->hw_bogomips, "?");
	else
	    sprintf(hw->hw_bogomips, "%0.2f", bogomips);

	if (tempMHz == 0.0)
		sprintf(hw->hw_megahertz, "?");
	else
		sprintf(hw->hw_megahertz, "%d", (int)tempMHz);
}

#endif /* system_ia64 */


linuxinfo_mips.c
/*
        linuxinfo_mips.c

        Copyright (C) 2004-2006
        All Rights Reserved.

        Helge Kreutzmann <debian@helgefjell.de>,
	based on linuxinfo_m68k.c from Alex Buell <alex.buell@munted.eu>

        Version Author  Date            Comments
        ----------------------------------------------------------------------
        1.0.0   KRE     20040522        Initial development
        1.0.1   KRE     20051118        Include <string.h>
	1.0.2   KRE     20060309        Fix detection if no endian info is given
	1.0.3   KRE     20060409        Do not look for endianess

        This is the mips port of linuxinfo
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "linuxinfo.h"

#ifdef system_mips

void GetHardwareInfo(int fd, struct hw_stat *hw)
{
	int processors = 0; 
	float bogomips = 0.0;
	float tempbogo;

	char temp_string[BUFSIZ], temp_string2[BUFSIZ];
	char chip[BUFSIZ] = "Unknown";
	char endian[BUFSIZ]="Unknown";
	char fpu[BUFSIZ]="FPU Unknown";
	char fullchip[BUFSIZ];
	char *pos;

	FILE *procfile;

	while (read_line(fd, temp_string, BUFSIZ) != 0)
	{
		splitstring(temp_string, temp_string2);

		if (processors == 0)
                       if (!(strncmp(temp_string, "cpu model", strlen("cpu model"))))
			{
				pos=strstr(temp_string2, "FPU");
				if (pos)
				{
				    strcpy(fpu, pos);
				    while (isspace(*(pos-1))) pos--;
				    *(pos) = '\0';
				}
				strcpy(chip, temp_string2);
			}

//              if (!(strncmp(temp_string, "byteorder", strlen("byteorder"))))
//			strcpy(endian, temp_string2);

                if (!(strncmp(temp_string, "BogoMIPS", strlen("BogoMIPS"))))
                {
                        processors++;
                        sscanf(temp_string2, "%f", &tempbogo);
                        bogomips += tempbogo;
                }
	}

	sprintf(hw->hw_memory, LONGSPEC, getphysicalmemory());
	hw->hw_processors = processors;

	if (fpu)
	{
		sprintf(fullchip, "%s (%s)",chip,fpu);
	}
	else
	{
		strcpy(fullchip,chip);
	}
	if (strncmp(endian, "Unknown", strlen("Unknown")))
	{
		sprintf(hw->hw_cpuinfo, "%s %s", endian,fullchip);
	}
	else
	{
		sprintf(hw->hw_cpuinfo, "%s", fullchip);
	}

	if (bogomips == 0.0)
	    sprintf(hw->hw_bogomips, "?");
	else
	    sprintf(hw->hw_bogomips, "%0.2f", bogomips);

	sprintf(hw->hw_megahertz, "?");
}
#endif /* system_mips */


linuxinfo_intel.c
/*
        linuxinfo_intel.c

        Copyright (C) 1998-2000
        All Rights Reserved.

        Alex Buell <alex.buell@munted.eu>

	Copyright (C) 2003 
	Brian Almeida <bma@debian.org>

	Copyright (C) 2004
	Martin Michlmayr <tbm@cyrius.com>

	Copyright (C) 2004-2007,2009,2010,2013 
	Helge Kreutzmann <debian@helgefjell.de>


        Version Author  Date            Comments
        ----------------------------------------------------------------------
        1.0.0   AIB     199803??        Initial development
	1.0.1	AIB	20000405	Updated and reworked code
	1.0.2	AIB	20000527	Added AMD K7
	1.0.3	AIB	20000815	Added AMD K6-3 500MHz
	1.0.4	AIB	20010909	Added getphysicalmemory() call
	1.0.5   OPAL    20030417        Added Athlon XP and MP.
	1.0.6   KRE     20040711        Added Opteron
	1.0.7   KRE     20040726        Added Pentium M
	1.0.8   KRE     20040809        Add further machines (Athlons, Xeons,
					Pentium 4s, Celerons), fix detection of
					Pentium MMX
	1.0.9   KRE     20050215        Added several amd64 machines
	1.0.10  KRE     20051106        Added another Pentium4 machine
	1.0.11  KRE     20051113        Added another Celeron machine
	1.0.12  KRE     20051120        Include <string.h>
	1.0.13  KRE     20060212        Add an AMP sempron cpu
	1.0.14  KRE     20060423        Add another AMD Sempron
	1.0.15  KRE     20060724        One more am64 and one pentium M added
	1.0.16  KRE     20070716        Added a Geode CPU
	1.0.17  KRE     20070810        Added another Xeon
	1.0.18  KRE     20080203        Added Core 2 and one more Pentium 4
	1.0.19  KRE     20090318        Added Core Solo
	1.0.20  KRE     20090524        Add another Xeon
	1.0.21  KRE     20090605        Add another Core Duo
	1.0.22  KRE     20100228        Add another Core Duo
	1.0.23  KRE     20110725        Add a Phenom
	1.0.24  KRE     20111104        Add another Xeon
	1.0.25  KRE     20130527        Add an Athlon X2
	1.0.26  KRE     20131114        Add another Xeon

        This is the x86 port of linuxinfo
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "linuxinfo.h"

#ifdef system_intel

static char *vendors[] =
{
	"GenuineIntel", "Intel",
	"AuthenticAMD", "AMD",
	"CyrixInstead", "Cyrix",
	"UMC UMC UMC ", "UMC",
	"CentaurHauls", "Centaur",
	"NexGenDriven", "NexGen",
	NULL, NULL
};

static char *models[] =
{
        "K6 (166 - 266)", "K6",
        "AMD-K6tm w/ mul", "K6 (MMX)",
        "K6-2 (PR233 - PR333)", "K6-2",
        "AMD-K6(tm) 3D", "K6-3 (3DNow)",
        "AMD-K6(tm) 3D+", "K6-3 (3DNow+)",
	"AMD-K6(tm)-III", "K6-3 (3DNowExt)",
	"AMD-K7(tm)", "K7",
	"Athlon", "Athlon",
	"AMD Athlon(tm) processor", "Athlon",
        "AMD Athlon(tm) Processor", "Athlon",
        "AMD Athlon(TM) XP", "Athlon XP",
        "AMD Athlon(tm) XP 1600+", "Athlon XP",
        "AMD Athlon(TM) MP", "Athlon MP",
	"AMD Opteron(tm) Processor 240", "Opteron",
	"AMD Opteron(tm) Processor 242", "Opteron",
	"AMD Opteron(tm) Processor 244", "Opteron",
	"AMD Opteron(tm) Processor 246", "Opteron",
	"AMD Opteron(tm) Processor 248", "Opteron",
	"AMD Sempron(tm)   2300+", "Sempron",
	"AMD Sempron(tm)   2600+", "Sempron",
	"AMD Athlon(tm) 64 Processor 2800+", "x86_64",
	"AMD Athlon(tm) 64 Processor 3000+", "x86_64",
	"AMD Athlon(tm) 64 Processor 3200+", "x86_64",
	"AMD Athlon(tm) 64 Processor 3400+", "x86_64",
	"AMD Athlon(tm) 64 Processor 3500+", "x86_64",
	"AMD Athlon(tm) 64 Processor 3700+", "x86_64",
	"AMD Athlon(tm) 64 X2 Dual Core Processor 5400+", "x86_64",
	"AMD Phenom(tm) II X6 1055T Processor", "Phenom II",
	"AMD A6-6310 APU with AMD Radeon R4 Graphics", "A6-6310 with Radeon R4",
        "Am5x86", "5x86",
        "C6", "C6",
	"C6-2", "C6-2",
        "6x86", "6x86",
        "6x86MX", "6x86MX",
        "Geode(TM) Integrated Processor by National Semi", "Geode by Nat. Sem.",
	"Geode(TM) Integrated Processor by AMD PCS", "Geode by AMD PCS",
        "Pentium 75+", "Pentium",
        "Pentium 75 - 200", "Pentium",
        "Pentium MMX", "Pentium MMX",
        "Mobile Pentium MMX", "Mobile Pentium MMX",
        "Pentium Pro", "Pentium Pro",
        "Pentium II (Klamath)", "Pentium II (Klamath)",
        "Pentium II (Deschutes)", "Pentium II (Deschutes)",
        "Pentium III (Katmai)", "Pentium III (Katmai)",
	"Pentium III (Coppermine)", "Pentium III (Coppermine)",
	"Intel(R) Pentium(R) 4 CPU 1.80GHz", "Pentium 4",
	"Intel(R) Pentium(R) 4 CPU 2.00GHz", "Pentium 4",
	"Intel(R) Pentium(R) 4 CPU 2.80GHz", "Pentium 4",
	"Intel(R) Pentium(R) 4 CPU 3.00GHz", "Pentium 4",
	"Intel(R) Pentium(R) 4 CPU 3.20GHz", "Pentium 4",
	"Intel(R) Pentium(R) M processor", "Pentium M",
	"Intel(R) Pentium(R) M processor 1600MHz", "Pentium M",
	"Intel(R) XEON(TM) CPU 2.20GHz", "Xeon",
	"Intel(R) XEON(TM) CPU 2.40GHz", "Xeon",
	"Intel(R) Xeon(TM) CPU 3.00GHz", "Xeon",
	"Intel(R) Xeon(TM) CPU 3.20GHz", "Xeon",
	"Intel(R) Xeon(TM) CPU 3.40GHz", "Xeon",
	"Intel(R) Xeon(TM) CPU 3.60GHz", "Xeon",
	"Intel(R) Xeon(R) CPU           L5420  @ 2.50GHz","Xeon",
        "Celeron (Coppermine)", "Celeron (Coppermine)",
        "Celeron (Mendocino)", "Celeron (Mendocino)",
        "Celeron (Covington)", "Celeron (Covington)",
	"Intel(R) Celeron(R) CPU 1.70GHz", "Celeron",
	"Intel(R) Celeron(R) CPU 2.20GHz", "Celeron",
	"Intel(R) Celeron(R) CPU 2.40GHz", "Celeron",
	"Intel(R) Celeron(R) M processor         1.40GHz", "Celeron M",
	"Mobile Intel(R) Pentium(R) 4     CPU 3.06GHz", "Pentium-M 4",
	"Intel(R) Core(TM) Solo CPU     U1400  @ 1.20GHz", "Core Solo",
	"Intel(R) Core(TM) i7-4600M CPU @ 2.90GHz", "Core i7",
	"Intel(R) Core(TM)2 CPU          6600  @ 2.40GHz", "Core 2",
	"Intel(R) Core(TM)2 Duo CPU     P8600  @ 2.40GHz", "Core 2 Duo",
	"Intel(R) Core(TM)2 Duo CPU     E7400  @ 2.80GHz", "Core 2 Duo",
	"Intel(R) Core(TM)2 Duo CPU     E8400  @ 3.00GHz", "Core 2 Duo",
	"Intel(R) Core(TM)2 Quad CPU    Q9550  @ 2.83GHz", "Core 2 Quad",
	"Genuine Intel(R) CPU            2140  @ 1.60GHz", "Pentium 4",
        NULL, NULL
};

void GetHardwareInfo(int fd, struct hw_stat *hw)
{
	char chip[BUFSIZ], vendor[BUFSIZ], model[BUFSIZ], Mhz[BUFSIZ], family[BUFSIZ];
	char temp_string[BUFSIZ], temp_string2[BUFSIZ];

	int processors = 0; 
	float bogomips = 0.0;
	float tempMHz = 0.0;

	char *p;

	strcpy(chip, "Unknown");
	strcpy(vendor, "Unknown");
	strcpy(model, "Unknown");
	strcpy(family, "Unknown");

	while (read_line(fd, temp_string, BUFSIZ) != 0) 
	{
		if (splitstring(temp_string, temp_string2))
		{
			if (strncmp(temp_string, "processor", strlen("processor")) == 0)
				processors++;

			if (strncmp(temp_string, "vendor_id", strlen("vendor_id")) == 0)
			{
				int i = 0;

				while (vendors[i] != NULL)
				{
					if (strncmp(temp_string2, vendors[i], strlen(vendors[i])) == 0)
						strcpy(vendor, vendors[++i]);

					i += 2;
				}
			}

			if (strncmp(temp_string, "model", strlen("model")) == 0) 
			{
				int i = 0;
				int found = 0;
		
				while ((models[i] != NULL)&&(!found))
				{
					if (strncmp(temp_string2, models[i], strlen(models[i])) == 0)
					{
						strcpy(model, models[++i]);
						found=1;
					}
					i += 2;
				}
			}

			if (strncmp(temp_string, "cpu family", strlen("cpu family")) == 0)
				strcpy(family, temp_string2);

			if (strncmp(temp_string, "cpu MHz", strlen("cpu MHz")) == 0)
				tempMHz = atol(temp_string2);

			if (strncmp(temp_string, "bogomips", strlen("bogomips")) == 0)
				bogomips = bogomips + atof(temp_string2);
		}

		
	}

	sprintf(hw->hw_memory, LONGSPEC, getphysicalmemory());
	hw->hw_processors = processors;
	sprintf(hw->hw_cpuinfo, "%s %s", vendor, model);

	if ((strcmp(vendor, "Unknown") == 0) && (strcmp(model, "Unknown") == 0))
	{
		switch (family[0])
		{
			case '4': sprintf(hw->hw_cpuinfo, "Unknown 486");
				break;
		}
	}

	if (bogomips == 0.0)
		sprintf(hw->hw_bogomips, "?");
	else
		sprintf(hw->hw_bogomips, "%0.2f", bogomips);

	if (tempMHz == 0.0)
		sprintf(hw->hw_megahertz, "?");
	else
		sprintf(hw->hw_megahertz, "%d", (int)tempMHz);
}

#endif /* system_intel */
